### PR TITLE
Saas 14.5 product partner wan

### PR DESCRIPTION
**Pull request by william-andre**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/fields.py [hunk 1]
```diff
@@ -2444,6 +2444,12 @@ def _setup_attrs(self, model_class, name):
...
                         continue
+                    if val.startswith('set '):
+                        assert val[4:] in dict(self.selection), (
+                            "%s: ondelete policy of type 'set %%' must be either 'set null', "
+                            "'set default', or 'set value' where value is a valid selection value."
+                        ) % self
+                        continue
                     raise ValueError(
                         "%r: ondelete policy %r for selection value %r is not a valid ondelete "
                         "policy, please choose one of 'set null', 'set default', 'cascade' or "
```
Comment (qdp-odoo) [inline odoo/fields.py:2455]: complete this message with the new possibility

**Review by qdp-odoo — APPROVED** [review]
I think the commit messages (especially the one touching models.py and fields.py) need to be improved.

Apart from that, it looks good to me. Maybe someone from @odoo/rd-framework-py can give his insight?

File: odoo/fields.py (hunk 1 above)
Comment (william-andre) [reply odoo/fields.py]: Done, with a more complete commit message

File: odoo/fields.py [hunk 2]
```diff
@@ -2423,32 +2423,37 @@ def _setup_attrs(self, model_class, name):
...
                             "field, or change the chosen ondelete policy" % self
                         )
-                        continue
-                    raise ValueError(
-                        "%r: ondelete policy %r for selection value %r is not a valid ondelete "
-                        "policy, please choose one of 'set null', 'set default', 'cascade' or "
-                        "a callable" % (self, val, key)
-                    )
+                    elif val.startswith('set '):
+                        assert val[4:] in dict(self.selection), (
```
Comment (rco-odoo) [inline odoo/fields.py:2447]: At that point, `self.selection` is actually not valid
```suggestion
                        assert val[4:] in values, (
```

**Review by rco-odoo — CHANGES_REQUESTED** [review]
Don't modify the ORM without writing some test cases.
See `TestSelectionOndelete` and `TestSelectionOndeleteAdvanced` in module `test_new_api`.

**Comment by rco-odoo** [comment]
@william-andre I just pushed a number fixup commits.
I reverted your change of a `ValueError` to an `AssertionError`, because we actually wanted that error.
I also improved the tests to cover the code you added in `ir.model`, and removed useless cases.
I also adapted the docstring of selection fields.

I let you review and squash them.

**Comment by william-andre** [comment]
>  I also improved the tests to cover the code you added in `ir.model`, and removed useless cases.

I feel like testing that things work when they should is as important as testing that they don't work when they shouldn't. But ok.

**Comment by rco-odoo** [comment]
> > I also improved the tests to cover the code you added in `ir.model`, and removed useless cases.
> 
> I feel like testing that things work when they should is as important as testing that they don't work when they shouldn't. But ok.

The test cases I remove are actually covered by the test models (in module `test_new_api`) themselves.
If they weren't correct, you wouldn't be able to install the module, in fact.
That's why the test class isn't covering the "correct" cases.

**Review by rco-odoo — APPROVED** [review]
The framework part looks good to me.

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by william-andre** [comment]
@robodoo retry

**Comment by william-andre** [comment]
@robodoo retry

**Comment by qdp-odoo** [comment]
@robodoo r+

File: addons/product/models/product_template.py [hunk 3]
```diff
@@ -417,12 +416,13 @@ def _compute_type(self):
...
-    # def _constrains_detailed_type(self):
-    #     type_mapping = self._detailed_type_mapping()
-    #     for record in self:
-    #         if record.type != type_mapping.get(record.detailed_type, record.detailed_type):
-    #             raise ValidationError(_("The Type of this product doesn't match the Detailed Type"))
+    @api.constrains('type', 'detailed_type')
+    def _constrains_detailed_type(self):
+        type_mapping = self._detailed_type_mapping()
+        for record in self:
+            print(record.type, record.detailed_type)
```
Comment (JKE-be) [inline addons/product/models/product_template.py:423]: :roll_eyes:

Comment (william-andre) [reply addons/product/models/product_template.py:423]: You didn't see nothing

**Pull request closed without merging**